Add `revealable()` API to `TextInput`

Stichoza opened the pull request:
- [x] Changes have been thoroughly tested to not break existing functionality.
- [x] New functionality has been documented or existing documentation has been updated to reflect changes.
- [x] Visual changes are explained in the PR description using a screenshot/recording of before and after.

## Description

This PR adds a `revealable()` method on `TextField` that will make password fields revealable. An eye icon will be displayed at the end of the input and clicking on it will toggle input type between `password` and `text`.

```php
use Filament\Forms\Components\TextInput;

TextInput::make('api_key')
    ->password()
    ->revealable()
```

It is also possible to pass second parameter to `password()` method instead:

```php
use Filament\Forms\Components\TextInput;

TextInput::make('api_key')
    ->password(revealable: true)
```

## Screenshots

Revealable field in default state:

<img width="674" alt="SCR-20230806-go" src="https://github.com/filamentphp/filament/assets/1139050/f02df7b0-8d07-4a8a-974e-237b83d333ea">


Revealable field after clicking on the 👁️ icon:

<img width="674" alt="SCR-20230806-gx" src="https://github.com/filamentphp/filament/assets/1139050/85b542ba-090a-4c3f-8cb6-3045a0dba3c7">

Suffixed fields are also supported:

<img width="531" alt="Screenshot 2023-08-06 at 00 35 25" src="https://github.com/filamentphp/filament/assets/1139050/94f33223-b02f-4401-a5b8-0abc2ba5b6b4">



> **Note**
> Calling `revealable()` on non-password fields will have no effect.

Stichoza committed:
Implement revealable functionality for password inputs.

Stichoza committed:
Align icons and hide slash icon with `x-cloak` to avoid visibility on load.

Stichoza committed:
Rebuild theme to include additional classes used in reveal button.

Stichoza committed:
Support inputs with suffixes.

Stichoza committed:
Add documentation explaining usage of revealable input.

aeq-dev commented:
@Stichoza is it available on table column?

Stichoza commented:
@aeq-dev No, this PR adds a feature for `TextInput`, but that's a really good idea for another pull request.

Stichoza committed:
Merge branch '3.x' into feature/revealable

Stichoza commented:
Merged 3.x and rebuilt css to resolve conflict in themes.css

Stichoza committed:
Add aliases for revealable icons

Stichoza committed:
Document icon aliases for revealable input.

Stichoza committed:
Remove classes and absolutely positioned link

Stichoza committed:
Add icon button with alias and simplified action

Stichoza committed:
Match color with other input actions.

Stichoza committed:
Merge branch '3.x' into feature/revealable

Stichoza commented:
@zepfietje, Done:
 - [x] Use an icon button for the actions.
 - [x] Place the actions next to the other actions to ensure it works nicely with inline or non-inline suffixes.
 - [x] Add aliases for the new icons.

### Screenshots

<details>
<summary>Regular</summary>
<img width="680" alt="SCR-20230813-6a0" src="https://github.com/filamentphp/filament/assets/1139050/918a0d96-43fa-483f-a598-387864857979">
</details>

<details>
<summary>Revealed</summary>
<img width="680" alt="SCR-20230813-6ai" src="https://github.com/filamentphp/filament/assets/1139050/ca1d5456-640a-406b-930f-227f730ad453">
</details>

<details>
<summary>With suffix</summary>
<img width="680" alt="SCR-20230813-6at" src="https://github.com/filamentphp/filament/assets/1139050/96a093d1-3d80-429a-aced-b9f77c6a4456">
</details>

<details>
<summary>Inline suffix with icon</summary>
<img width="680" alt="SCR-20230813-6bb" src="https://github.com/filamentphp/filament/assets/1139050/59ff4576-2e56-4899-83c3-fbaae09a8fad">
</details>

<details>
<summary>Suffix action</summary>
<img width="680" alt="SCR-20230813-6dn" src="https://github.com/filamentphp/filament/assets/1139050/e214a272-c4aa-428a-85b7-99dda716dae4">
</details>


The icon has no border between input (similar to inline suffix) and that is intentional because the "show password" buttons are usually drawn like that in major operating systems.

Stichoza committed:
Revert: Support inputs with suffixes.

zepfietje commented:
Thanks, @Stichoza.

Maybe I wasn't clear, but I think it's better if we pass the reveal actions to the actual suffix actions slot so they're grouped inside the border with the other actions when not using inline suffix.

Could you handle that final thing? 🙂

Stichoza committed:
Group revealable action with suffix actions.

Stichoza commented:
Ok, sure. I just thought it looked better outside the border. I changed the way you suggested and now it looks like this:

<img width="1039" alt="image" src="https://github.com/filamentphp/filament/assets/1139050/8b4730ad-cc0f-4599-94fd-0470ccbe0a42">

It's still possible to have reveal action without border using `->revealable()->inlineSuffix()` combination so it's flexible in that way.

Check this out and let me know if I should change anything 🙌🏼

Stichoza committed:
Move icon after suffix text

Stichoza committed:
Fix formatting

zepfietje commented on packages/forms/resources/views/components/text-input.blade.php:
```diff
@@ -28,11 +29,13 @@
         :suffix-actions="$suffixActions"
         :suffix-icon="$suffixIcon"
         :valid="! $errors->has($statePath)"
+        :is-revealable="$isRevealable"
```
For Blade component attributes we follow naming like standard HTML elements, so could you change this to revealable?

zepfietje commented on packages/forms/resources/views/components/text-input.blade.php:
```diff
@@ -28,11 +29,13 @@
             \Filament\Support\prepare_inherited_attributes($getExtraAttributeBag())
                 ->class(['overflow-hidden'])
         "
+        x-data="{inputType: '{{ $getType() }}'}"
```
I think we should only render this when revealable.

Stichoza replied:
Is rendering by `\Illuminate\Support\Js` be okay? Like this:

```php
:x-data="$isRevealable ? \Illuminate\Support\Js::from(['inputType' => $getType()]) : null"
```

zepfietje replied:
We don't need it to be an Alpine component at all when not revealable.

Stichoza replied:
Yes, it doesn't render as Alpine component, `x-data` is not present at all this way

zepfietje commented on packages/forms/resources/views/components/text-input.blade.php:
```diff
@@ -60,6 +63,7 @@ class="fi-fo-text-input"
                         $applyStateBindingModifiers('wire:model') => $statePath,
                         'x-data' => (count($extraAlpineAttributes) || filled($mask)) ? '{}' : null,
                         'x-mask' . ($mask instanceof \Filament\Support\RawJs ? ':dynamic' : '') => filled($mask) ? $mask : null,
+                        'x-bind:type' => 'inputType',
```
Should only be added when revealable.

zepfietje commented on packages/support/resources/views/components/input/wrapper.blade.php:
```diff
@@ -174,6 +175,26 @@
             @endif
 
+            @if ($isRevealable)
+                <div @class([
+                    $actionsClasses,
+                    'select-none'
+                ]) >
+                    <x-filament::icon-button
+                        icon-alias="forms::components.text-input.show-password"
+                        icon="heroicon-o-eye"
+                        color="gray"
+                        x-on:click="inputType = 'text'"
+                        x-show="inputType === 'password'"/>
+                    <x-filament::icon-button
+                        icon-alias="forms::components.text-input.hide-password"
+                        icon="heroicon-o-eye-slash"
+                        color="gray"
+                        x-on:click="inputType = 'password'"
+                        x-show="inputType !== 'password'" x-cloak/>
+                </div>
+            @endif
```
Could you add a blank line between elements, sort attributes alphabetically and fix formatting issues?

Stichoza replied:
Blank line just between icon-buttons or also between parent div and icons?

zepfietje replied:
Generally, we always add a blank line just between sibling blocks. 🙂

Stichoza replied:
Roger that 🫡

Stichoza committed:
Match Blade component attribute naming standard

Stichoza committed:
Render `x-data` and `x-bind:type` only when revealable

Stichoza committed:
Remove redundant `select-none` class

Element is now button instead of a link, so `select-none` class is no longer relevant.

Stichoza committed:
Add spaces between elements, sort attributes, fix formatting

Stichoza committed:
Merge branch '3.x' into feature/revealable

Stichoza committed:
Fix class attribute

Stichoza committed:
Spacing sibling elements only

zepfietje commented on packages/support/resources/views/components/input/wrapper.blade.php:
```diff
@@ -174,6 +175,27 @@
+
             @if (count($suffixActions))
                 <div class="{{ $actionsClasses }}">
                     @foreach ($suffixActions as $suffixAction)
```
Shouldn't the actions actually be moved to here, @Stichoza? 🤔

Stichoza replied:
Correct, changed.

zepfietje reviewed (changes requested):
Great job on the changes, @Stichoza!
I probably didn't see this before, but have a look at my comment. 🙂

zepfietje commented on packages/forms/resources/views/components/text-input.blade.php:
```diff
@@ -60,6 +63,7 @@ class="fi-fo-text-input"
                         $applyStateBindingModifiers('wire:model') => $statePath,
                         'x-data' => (count($extraAlpineAttributes) || filled($mask)) ? '{}' : null,
```
Sorry for missing this initially, but the `x-data` in here should be used instead of the one added in this PR?

Stichoza replied:
In that case the icon button won't be able to change the state of `inputType` variable as it is outside the text input scope (it's inside the wrapper, not a child of input component). The `x-show` of icons also depends on the `inputType` variable.

zepfietje replied:
Whoops, sorry, the collapsed diff confused me here. 😅

Stichoza committed:
Move icons inside the suffix actions wrapper div

pxlrbt commented:
A bit late to the party, but I am in favor of using `->password(revealable: true)` and dropping `->revealable()` since it's only working on password fields anyway.

ziming commented:
it can be for api keys.etc as well I think

Stichoza commented:
@pxlrbt Yeah, that's right. But I added `->revealable()` to make code look more easy to read and match other available methods. Similar to current `->numeric()->step(10)` vs `->numeric(step: 10)`.

@ziming yes, api keys and etc, will still be password inputs just with different name.

Stichoza commented:
@zepfietje Are there any more changes needed? Should I remove `->revealable()` and leave only `->password(revealable: true)`?

zepfietje commented:
Let's see what @danharrin thinks about the API.
Then I'll do a final review afterwards.

caendesilva commented:
Would be really great to have included on the default login page. It also improves accessibility so I am definitely in favour of this PR!

danharrin commented:
Hi, thanks for the PR. I think it's important to give you a bit of information as to why this PR has been tagged as "low priority". We really appreciate all contributions, and intend to visit this in the future, and hopefully merge it if it is a suitable feature for the framework.

However, reviewing PRs, and often rewriting them on behalf of the author to ensure they are consistent with the rest of the framework, consumes a lot of my time. At the moment we have a large backlog of bug reports that I am prioritising, since most of these affect functionality that worked in a previous version and are now impacting real projects. Hopefully you can see how that work comes first before introducing new features that people aren't using in projects yet.

Again, thank you for contributing and I'm looking forward to reviewing this PR when our backlog of bug reports has reduced.

binaryfire commented:
It'd be great to have this on the Textarea field as well. Things like TLS certificates, SSH keys etc are multi-line

saade commented:
Maybe `->secret(isRevealable: true)` ?

Then `->password()` sets it to true, and can be turned off with `->revealable(false)`

zepfietje commented on packages/support/resources/views/components/input/wrapper.blade.php:
```diff
@@ -174,11 +175,32 @@
-                        {{ $suffixAction }}
-                    @endforeach
+                    @if ($revealable)
+                        <x-filament::icon-button
```
@danharrin, what do you think about moving these to action methods like we have for repeater/builder?

danharrin replied:
Potentially this should consume the suffix action API

zepfietje replied:
How? That would only be possible if we could trigger JS code from the suffix action, right?

danharrin replied:
We trigger JS code from actions for collapse/expand on repeaters etc

zepfietje replied:
Yeah I know. Could you handle reworking the back-end side of this PR then?

danharrin replied:
Yeah I will do

danharrin commented:
Thanks for your work on this! I've rewritten it using the latest Filament version in #10326. It will be released as a feature of v3.2.

Stichoza commented:
Thanks for adding the feature! 🎉  A bit sad that my contribution won't show up anywhere 😢

danharrin commented:
Ill try and remember to mention you in the changelog when it gets written for v3.2

danharrin commented:
You've been credited in https://github.com/filamentphp/filament/releases/tag/v3.2.0 😄
